### PR TITLE
Stop returning "es." internal exception headers as http response headers

### LINKED ISSUE #17593: Stop returning additional HTTP headers on exceptions
When an `ElasticsearchException` is bubbled up to HTTP, there are additional headers being returned.

``` bash
# curl -X DELETE localhost:9200/test
{"acknowledged":true}

# curl -X PUT localhost:9200/test -v
*   Trying 127.0.0.1...
* Connected to localhost (127.0.0.1) port 9200 (#0)
> PUT /test HTTP/1.1
> Host: localhost:9200
> User-Agent: curl/7.43.0
> Accept: */*
>
< HTTP/1.1 200 OK
< Content-Type: application/json; charset=UTF-8
< Content-Length: 21
<
* Connection #0 to host localhost left intact
{"acknowledged":true}


# curl -X PUT localhost:9200/test -v
*   Trying 127.0.0.1...
* Connected to localhost (127.0.0.1) port 9200 (#0)
> PUT /test HTTP/1.1
> Host: localhost:9200
> User-Agent: curl/7.43.0
> Accept: */*
>
< HTTP/1.1 400 Bad Request
< es.index: test
< Content-Type: application/json; charset=UTF-8
< Content-Length: 203
<
* Connection #0 to host localhost left intact
{"error":{"root_cause":[{"type":"index_already_exists_exception","reason":"already exists","index":"test"}],"type":"index_already_exists_exception","reason":"already exists","index":"test"},"status":400}
```

If those headers are not needed on HTTP, maybe it does not make any sense to return them? All the information looks available to me in the HTTP response body, but I am possibly missing some cases.

### PR DESCRIPTION
**Pull request by javanna**
Move "es." internal headers to separate metadata set in `ElasticsearchException` and stop returning them as response headers. These headers are printed out as part of the response body already, they weren't meant to be sent back as response headers too. They were introduced to prevent having to add custom exceptions to our codebase whenever we need to throw an exception that has to hold some additional metadata that `ElasticsearchException` doesn't support out of the box. The header notion stays in `ElasticsearchException` but only for what actually needs to be returned as response header (no "es." prefix).

Also removed `ESExceptionTests` and moved its tests under `ElasticsearchExceptionTests` or `ExceptionSerializationTests`

Closes #17593

### CODE REVIEW

**Review by tlrx — COMMENTED** [review]
Thanks @javanna, I think this is heading in the right direction. I left some minor comments but we still need to figure out what to do with the TODOs :)

I do like the move of internal headers es.* to exception metadata and I share your concerns about serializing back a parsed exception... but that shouldn't happen, neither in core nor in the high level rest client.

For a long term, I think we should also drop the metadataToXcontent() method which can lead to name conflicts in the rendered xcontent and also adds confusion if we add a metadata class attribute.

**Comment by javanna** [comment]
> For a long term, I think we should also drop the metadataToXcontent() method which can lead to name conflicts in the rendered xcontent and also adds confusion if we add a metadata class attribute.

that's correct, I think that the problem is that those additional fields are printed at the same level as the metadata (previous es. headers). They should rather be in a different object that only subclasses can write, but that is a breaking change. That said I am not sure how many clients are able to parse our exceptions entirely given how difficult we made it for them :)

File: core/src/main/java/org/elasticsearch/ElasticsearchException.java [hunk 1]
```diff
@@ -410,6 +470,8 @@ public static ElasticsearchException fromXContent(XContentParser parser) throws
 
         String type = null, reason = null, stack = null;
         ElasticsearchException cause = null;
+        Map<String, List<String>> metadata = new HashMap<>();
+        //TODO should this be a Map<String, List<String>> instead?
         Map<String, Object> headers = new HashMap<>();
```
Comment (javanna) [inline core/src/main/java/org/elasticsearch/ElasticsearchException.java]: ok I will remove the TODO then

File: core/src/main/java/org/elasticsearch/ElasticsearchException.java [hunk 2]
```diff
@@ -423,8 +485,8 @@ public static ElasticsearchException fromXContent(XContentParser parser) throws
                 } else if (STACK_TRACE.equals(currentFieldName)) {
                     stack = parser.text();
                 } else {
-                    // Everything else is considered as a header
-                    headers.put(currentFieldName, parser.text());
+                    //TODO here we need to parse multiple values
+                    metadata.put(currentFieldName, Collections.singletonList(parser.text()));
```
Comment (javanna) [inline core/src/main/java/org/elasticsearch/ElasticsearchException.java]: Are we testing this somewhere?

File: core/src/test/java/org/elasticsearch/rest/BytesRestResponseTests.java [hunk 3]
```diff
@@ -45,18 +45,18 @@
...
-
     }
 
     public void testWithHeaders() throws Exception {
         RestRequest request = new FakeRestRequest();
         RestChannel channel = randomBoolean() ? new DetailedExceptionRestChannel(request) : new SimpleExceptionRestChannel(request);
 
         BytesRestResponse response = new BytesRestResponse(channel, new WithHeadersException());
+        assertEquals(2, response.getHeaders().size());
+        assertThat(response.getHeaders().get("n1"), notNullValue());
```
Comment (javanna) [inline core/src/test/java/org/elasticsearch/rest/BytesRestResponseTests.java]: good catch

**Comment by javanna** [comment]
@s1monw mind having a look too?

**Review by s1monw — APPROVED** [review]
LGTM

**Review by jaymode — APPROVED** [review]

**Pull request merged**